react-components/src/components/RuleBasedOutputs/utils.ts

```diff
@@ -289,6 +413,68 @@ const analyzeNodesAgainstExpression = async ({
   return applyNodeStyles(filteredAllTreeNodes, outputSelected, model);
 };
 
+const generateTimeseriesAndDatapointsFromTheAsset = ({
+  contextualizedAssetNode,
+  assetIdsAndTimeseries,
+  timeseriesDatapoints
+}: {
+  contextualizedAssetNode: Asset;
+  assetIdsAndTimeseries: AssetIdsAndTimeseries[];
+  timeseriesDatapoints: Datapoints[] | undefined;
+}): TimeseriesAndDatapoints[] => {
+  const timeseriesLinkedToThisAsset = assetIdsAndTimeseries.filter(
+    (item) => item.assetIds?.externalId === contextualizedAssetNode.externalId
+  );
+
+  const timeseries = timeseriesLinkedToThisAsset?.map((item) => item.timeseries).filter(isDefined);
+  const datapoints = timeseriesDatapoints?.filter((datapoint) =>
+    timeseries?.find((item) => item?.externalId === datapoint.externalId)
+  );
+
+  const timeseriesData: TimeseriesAndDatapoints[] = timeseries
+    .map((item) => {
+      const datapoint = datapoints?.find((datapoint) => datapoint.externalId === item.externalId);
+      if (datapoint === undefined) return undefined;
+
+      const content: TimeseriesAndDatapoints = {
+        ...item,
+        ...datapoint
+      };
+      return content;
+    })
+    .filter(isDefined);
+  return timeseriesData;
+};
+
+export const traverseExpressionToGetTimeseries = (
+  expressions: Expression[] | undefined
+): string[] | undefined => {
+  let timeseriesExternalIdFound: string[] | undefined = [];
+
+  const timeseriesExternalIdResults = expressions
+    ?.map((expression) => {
+      switch (expression.type) {
+        case 'or':
+        case 'and': {
+          timeseriesExternalIdFound = traverseExpressionToGetTimeseries(expression.expressions);
+          break;
+        }
+        case 'not': {
+          timeseriesExternalIdFound = traverseExpressionToGetTimeseries([expression.expression]);
+          break;
+        }
+        case 'numericExpression': {
+          timeseriesExternalIdFound = getTimeseriesExternalIdFromNumericExpression(expression);
```

Comment:
Are we restricting ourselves to only use time series that have an `externalId`? Are we sure that won't cause us trouble?

Reply:
AFAIK all timeseries have to have an externalId, isnt?

Reply:
It seems like that's not the case. The API lists it as an optional field: https://api-docs.cognite.com/20230101/tag/Time-series/operation/getTimeSeries (look at the list result)

Reply:
as we've discussed in DM, let's skip the internalID due to the fact of relationships endpoints used to get timeseries through relationship support only externalId.